src/framework/linalg/linops/linops_aer_vector.hpp

```diff
@@ -0,0 +1,142 @@
+/**
+ * This code is part of Qiskit.
+ *
+ * (C) Copyright IBM 2018, 2019.
+ *
+ * This code is licensed under the Apache License, Version 2.0. You may
+ * obtain a copy of this license in the LICENSE.txt file in the root directory
+ * of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Any modifications or derivative works of this code must retain this
+ * copyright notice, and modified files need to carry a notice indicating
+ * that they have been altered from the originals.
+ */
+
+#ifndef _aer_framework_linalg_linops_aer_vector_hpp_
+#define _aer_framework_linalg_linops_aer_vector_hpp_
+
+#include <functional>
+#include <vector>
+
+#include "framework/linalg/almost_equal.hpp"
+#include "framework/linalg/vector.hpp"
+
+namespace AER {
+namespace Linalg {
+
+// This defines functions add, sub, mul, div and iadd, imul, isub, idiv
+// that work for numeric vector types
+
+//----------------------------------------------------------------------------
+// Linear operations
+//----------------------------------------------------------------------------
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T> add(const Vector<T>& lhs, const Vector<T>& rhs) {
+  return lhs + rhs;
+}
+
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T>& iadd(Vector<T>& lhs, const Vector<T>& rhs) {
+  lhs += rhs;
+  return lhs;
+}
+
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T> sub(const Vector<T>& lhs, const Vector<T>& rhs) {
+  return lhs - rhs;
+}
+
+template <class T, typename = enable_if_numeric_t<T>>
+Vector<T>& isub(Vector<T>& lhs, const Vector<T>& rhs) {
+  lhs -= rhs;
+  return lhs;
+}
+
+//----------------------------------------------------------------------------
+// Affine operations
+//----------------------------------------------------------------------------
+template <class T, class Scalar, typename = enable_if_numeric_t<T>,
+          typename = enable_if_numeric_t<Scalar>>
+Vector<T> add(const Vector<T>& data, const Scalar& val) {
+  Vector<T> result(data.size(), false);
+  const T cast_val(val);
+  std::transform(data.data(), data.data() + data.size(), result.data(),
+                 [&cast_val](const T& a)->T{ return a + cast_val; });
+  return result;
+}
+
+template <class T, class Scalar, typename = enable_if_numeric_t<T>,
+          typename = enable_if_numeric_t<Scalar>>
+Vector<T>& iadd(Vector<T>& data, const Scalar& val) {
+  const T cast_val(val);
+  std::for_each(data.data(), data.data() + data.size(),
+                [&cast_val](T& a)->T{ a += cast_val; });
+  return data;
+}
+
+template <class T, class Scalar, typename = enable_if_numeric_t<T>,
+          typename = enable_if_numeric_t<Scalar>>
+Vector<T> sub(const Vector<T>& data, const Scalar& val) {
+  Vector<T> result(data.size(), false);
```

Comment:
To avoid code duplication as well we can implement `sub` function in terms of `add`:
```
T sub(const T& vec, const Scalar& val){
    return add(vec, -val);
}
```